src/core/components/files-regular/add-async-iterator.js

```diff
@@ -0,0 +1,200 @@
+function validateInput () {
+  return async function * (source) {
+    for await (const data of source) {
+      validateAddInput(data)
+
+      yield data
+    }
+  }
+}
+
+function normalizeContent (content) {
+  if (supportsFileReader && kindOf(content) === 'file') {
+    return streamFromFileReader(content)
+  }
+
+  // pull stream source
+  if (isSource(content)) {
+    return toAsyncIterator(content)
+  }
+
+  if (typeof content === 'string') {
+    return Buffer.from(content)
+  }
+
+  if (Array.isArray(content) && content.length && !Array.isArray(content[0])) {
+    return [content]
+  }
+
+  return content
+}
+
+function normalizeInput (opts) {
+  return async function * (source) {
+    for await (let data of source) {
+      if (data.content) {
+        data.content = normalizeContent(data.content)
+      } else {
+        data = {
+          path: '',
+          content: normalizeContent(data)
+        }
+      }
+
+      if (opts.wrapWithDirectory && !data.path) {
+        throw new Error('Must provide a path when wrapping with a directory')
+      }
+
+      yield data
+    }
+  }
+}
+
+function doImport (ipfs, opts) {
+  return function (source) {
+    return importer(source, ipfs._ipld, opts)
+  }
+}
+
+function prepareFile (ipfs, opts) {
+  return async function * (source) {
+    for await (const file of source) {
+      let cid = file.cid
+      const b58Hash = cid.toBaseEncodedString()
+
+      if (opts.onlyHash) {
+        yield {
+          path: file.path === undefined ? b58Hash : (file.path || ''),
+          hash: b58Hash,
+          size: file.unixfs.fileSize()
+        }
+
+        return
+      }
+
+      const node = await ipfs.object.get(file.cid, Object.assign({}, opts, { preload: false }))
+
+      if (opts.cidVersion === 1) {
+        cid = cid.toV1()
+      }
+
+
+      let size = node.size
+
+      if (Buffer.isBuffer(node)) {
+        size = node.length
+      }
+
+      yield {
+        path: file.path === undefined ? b58Hash : (file.path || ''),
+        hash: b58Hash,
+        size
+      }
+    }
+  }
+}
+
+function preloadFile (ipfs, opts) {
+  return async function * (source) {
+    for await (const file of source) {
+      const isRootFile = !file.path || opts.wrapWithDirectory
+        ? file.path === ''
+        : !file.path.includes('/')
+
+      const shouldPreload = isRootFile && !opts.onlyHash && opts.preload !== false
+
+      if (shouldPreload) {
+        ipfs._preload(file.hash)
+      }
+
+      yield file
+    }
+  }
+}
+
+function pinFile (ipfs, opts) {
+  return async function * (source) {
+    for await (const file of source) {
+      // Pin a file if it is the root dir of a recursive add or the single file
+      // of a direct add.
+      const pin = 'pin' in opts ? opts.pin : true
+      const isRootDir = !file.path.includes('/')
+      const shouldPin = pin && isRootDir && !opts.onlyHash && !opts.hashAlg
+
+      if (shouldPin) {
+        await ipfs.pin.add(file.hash, { preload: false })
```

Comment:
We could probably do this asynchronously as well?

Reply:
Probably needs to be done in the gc read lock for add though